semtypes/src/main/java/io/ballerina/types/MappingAlternative.java

```diff
@@ -51,8 +51,10 @@ private MappingAlternative(SemType semType, MappingAtomicType[] pos, MappingAtom
                     BddPath.from());
             List<MappingAlternative> alts = new ArrayList<>();
             for (BddPath bddPath : paths) {
-                alts.add(from(cx, ComplexSemType.createComplexSemType(0,
-                        UniformSubtype.from(UniformTypeCode.UT_MAPPING_RW, bddPath.bdd)), bddPath.pos, bddPath.neg));
+                SemType semType = Core.createUniformSemType(UniformTypeCode.UT_MAPPING_RW, bddPath.bdd);
+                if (semType != PredefinedType.NEVER) {
```

Comment:
This might not work in some instances. Shall we use equals method here? We also need to implement equals method, at least on `UniformTypeBitSet`

Reply:
Can we simply use Java Object equals here, or should we create a method to check type and compare bitset and use here?

Reply:
Object.equals is just == right, so we have to implement the methods, I don't think we need to do that for complex semtypes as we really don't compare them with ==, if we do we have to implement them as well. 
As I see it we just need to implement UniformTypeBitSet

Reply:
if I'm not mistaken I believe `Object.equals` will compare the objects while `==` will check the reference. I'll make a comparator anyway so we can extend to complex semtypes in the future if needed.

Reply:
I think without overridding equals is the same as ==, my mistake.